[FLINK-23908][runtime] Hint user about possibly non-deterministic shu…

dmvk opened the pull request:
https://issues.apache.org/jira/browse/FLINK-23908

dmvk commented:
@flinkbot run azure

dmvk commented:
Failed with https://issues.apache.org/jira/browse/FLINK-22889

dmvk commented:
@flinkbot run azure

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyGroupRangeOffsets.java:
```diff
@@ -141,8 +150,7 @@ public KeyGroupRange getKeyGroupRange() {
         if (idx < 0 || idx >= offsets.length) {
-            throw new IllegalArgumentException(
-                    "Key group " + keyGroup + " is not in " + keyGroupRange + ".");
+            throw newIllegalKeyGroupException(keyGroup, keyGroupRange);
```
I think we are double throwing something here. Maybe let `newIllegalKeyGroupException` method only return the `IllegalStateException` instead of throwing it.

dmvk replied:
🤦 That was the original intention

tillrohrmann reviewed (approved):
Thanks for creating this PR @dmvk. The changes look good to me. I had a single minor comment. I think we should also port this fix back to `release-1.13` and `release-1.12`. Can you open PRs against these branches?

dmvk commented:
@tillrohrmann Thanks for the review! I've addressed your comments.

Backport PRs are available here:
- https://github.com/apache/flink/pull/16943 (1.12)
- https://github.com/apache/flink/pull/16942 (1.13)

tillrohrmann commented:
Great, thanks @dmvk. I'll merge them once Azure gives green light.
